rq/worker.py

```diff
@@ -516,20 +512,55 @@ def execute_job(self, job, queue):
         else:
             self._horse_pid = child_pid
             self.procline('Forked {0} at {1}'.format(child_pid, time.time()))
-            while True:
-                try:
-                    os.waitpid(child_pid, 0)
-                    self.set_state('idle')
-                    break
-                except OSError as e:
-                    # In case we encountered an OSError due to EINTR (which is
-                    # caused by a SIGINT or SIGTERM signal during
-                    # os.waitpid()), we simply ignore it and enter the next
-                    # iteration of the loop, waiting for the child to end.  In
-                    # any other case, this is some other unexpected OS error,
-                    # which we don't want to catch, so we re-raise those ones.
-                    if e.errno != errno.EINTR:
-                        raise
+
+    def monitor_work_horse(self, job):
+        """The worker will wait for the work horse and make sure it executes
+        within the given timeout bounds, or will end the work horse with
+        SIGALRM.
+        """
+        while True:
+            try:
+                _, ret_val = os.waitpid(self._horse_pid, 0)
```

Comment:
waitpid waits for a change in the state of the process (the child terminated; the child was stopped by a signal; or the child was resumed by a signal;). OOM sends a SIGKILL to the process which terminates it and a segfault is just a SIGSEGV to a process (the default handler terminates the process). Both cases should be handled by waitpid (with the exception where a custom signal handler is used for SIGSEGV, in which case I don't think we should intervene since the process handled whatever happened itself).